Examples/Nodify.Shapes/Canvas/CanvasViewModel.cs

```diff
@@ -0,0 +1,249 @@
+                {
+                    Name = $"User {i + 1}",
+                    Color = color,
+                    Location = new Point(_rand.Next(0, 1000), _rand.Next(0, 1000))
+                });
+            }
+
+            var ellipse = new EllipseViewModel
+            {
+                Location = new Point(100, 50),
+                Width = 150,
+                Height = 150
+            };
+            _shapes.Add(ellipse);
+
+            var rectangle = new RectangleViewModel
+            {
+                Location = new Point(400, 100),
+                Width = 150,
+                Height = 150
+            };
+            _shapes.Add(rectangle);
+
+            Connections.Add(new ConnectionViewModel(ellipse.RightConnector, rectangle.LeftConnector));
+
+            var ellipse2 = new EllipseViewModel
+            {
+                Location = new Point(100, 250),
+                Width = 150,
+                Height = 150
+            };
+            _shapes.Add(ellipse2);
+
+            var rectangle2 = new RectangleViewModel
+            {
+                Location = new Point(450, 400),
+                Width = 150,
+                Height = 150
+            };
+            _shapes.Add(rectangle2);
+
+            var triangle = new TriangleViewModel
+            {
+                Location = new Point(800, 200),
+                Width = 150,
+                Height = 150
+            };
+            _shapes.Add(triangle);
+
+            Connections.Add(new ConnectionViewModel(ellipse2.BottomConnector, rectangle2.TopConnector));
+            Connections.Add(new ConnectionViewModel(rectangle.RightConnector, rectangle2.RightConnector));
+
+            SelectedShapes.Add(triangle);
+
+            Decorators.Add(ShapeToolbar);
+            Decorators.AddRange(Cursors);
+
+            // Re-enable undo redo
+            UndoRedo.IsEnabled = true;
+        }
+
+        public void AddShape(ShapeViewModel shape)
+        {
+            var action = new DelegateAction(() => _shapes.Add(shape), () => _shapes.Remove(shape), "Add shape");
+            UndoRedo.ExecuteAction(action);
+        }
+
+        private void AddConnection(ConnectorViewModel source, ConnectorViewModel target)
+        {
+            var connection = new ConnectionViewModel(source, target);
+            var action = new DelegateAction(() => Connections.Add(connection), () => Connections.Remove(connection), "Connect");
+            UndoRedo.ExecuteAction(action);
+        }
+
+        private void RemoveConnection(ConnectionViewModel connection)
+        {
+            var action = new DelegateAction(() => Connections.Remove(connection), () => Connections.Add(connection), "Disconnect");
+            UndoRedo.ExecuteAction(action);
+        }
+
+        private bool CanConnect(ConnectorViewModel? source, ConnectorViewModel? target)
+        {
+            return source != null
+                && target != null
+                && source != target
```

Comment:
source != target could likely be ommited, should be possible to have repeating/recursive markup with connections for the sample.

Reply:
Removing the condition allows connecting connectors to themselves, which is not properly handled in any of the built-in connection types.

![image](https://github.com/miroiu/nodify/assets/12727904/d02cfa05-9d4b-4319-bde9-b47d9d03c0d1)